armsr: enable more features / hardware

mcbridematt opened the pull request:
This is an effort to bring the kernel config in armsr/armv8 closer to the Linux arm64 defconfig, so armsr/armv8 can run out of the box on as many platforms as possible.

Changes of note:

- Enable virtualization/KVM host
- Sync CPU errata options set with Linux defconfig
- Sync CPU features and scheduler configuration with Linux defconfig
- Fix NXP i.MX8 support (was 'damaged' in the process of updating to kernel 6.1, 5.15/23.05 is unaffected)
- Enable Renesas SoCs (RZ/G, RZ/V)
- Add Raspberry Pi 4 network support
This is *not* a replacement for `bcm27xx/bcm2711`, as it relies on an external EFI firmware. But it is useful to have for testing, especially when using an ACPI firmware.
- Added Marvell (former Cavium IP) ThunderX2 network driver. I missed this previously and assumed ThunderX and ThunderX2 had the same driver, it's different
- Package and install `kmod-mdio-thunder`. This was already being built (loosely selected by ThunderX) but not installed

mcbridematt committed:
armsr: armv8: sync Arm64 erratum options with kernel defconfig

To reduce differences with the Linux arm64 defconfig,
sync the enabled erratum items with defconfig.

There are still some options not selected due to
CONFIG_KVM or other options not enabled in OpenWrt
by default.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: fix NXP/Freescale i.MX family configuration

Due to an error on my part, Anton Antonov's
i.MX changes[1] did not fully make it into my
armvirt kernel 6.1 EFI pull request. I have updated
them using the options he supplied[1] as well
as comparing to the Linux arm64 defconfig.

The notable exception is:
CONFIG_USB_DWC3_OF_SIMPLE currently disabled
due to an issue with i.MX8P and i.MX8Q.

Fixes: 3efb3b8 ("armvirt: 64: Add NXP i.MX 8M Mini/Nano/Quad/Plus EVK support")

Signed-off-by: Mathew McBride <matt@traverse.com.au>

[1] - https://github.com/anta5010/openwrt/commit/ccf826c3444fd5fa8fac08aac05f66568ae0c242

mcbridematt committed:
armsr: armv8: enable KVM host

x86/64 enables support for KVM so I can't see a reason why
not on armsr/armv8 as well.

Arm CPU errata workaround items related to virtualization
are also enabled by this change.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: sync CPU features, EFI, CMA and scheduler options with Linux defconfig

To bring the armsr/armv8 kernel configuration closer to the Linux
arm64 defconfig, synchronize options related to CPU features
(especially more recent Armv8.X variants), scheduler, EFI vars,
CMA and scheduler options.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: enable CONFIG_ARCH_RENESAS

Renesas markets several embedded Arm64 SoCs in the
RZ series (RZ/G, RZ/V), so should be enabled in
a general purpose target.

Automotive (R-Car) SoC's are not enabled by this change.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: enable Broadcom arch'es

This is part of an effort to reduce differences between
the OpenWrt armsr/armv8 config and Linux arm64 defconfig.

This enables CONFIG_ARCH_BCM and downstream
CONFIG_ARCH_BCM2835 (= BCM2711 like Raspberry Pi 4)
and CONFIG_ARCH_BCM_IPROC (Broadcom iProc packet processors).

The broadband specific SoC's (ARCH_BCMBCA) are left out
as it is assumed these will not be doing EFI boot.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: add Broadcom GENET and MDIO modules

These are used by common Broadcom SoC's like
the BCM2711 (RPi4) and iProc network processor.

Tested on the RPi4B using the Raspberry Pi
UEFI+ACPI firmware[1].

Signed-off-by: Mathew McBride <matt@traverse.com.au>

[1] - https://github.com/pftf/RPi4

mcbridematt committed:
armsr: armv8: add bcmgenet (Raspberry Pi 4 GENET) to profile

kmod-bcmgenet is needed for Ethernet support on the
Raspberry Pi 4.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: add Marvell (Cavium) ThunderX2 network driver

The initial armv8 module incorrectly labelled the Thunder(v1) as
supporting the ThunderX2, when they have different drivers.

Add kmod-octeon-tx2 to support the newer devices.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: package and select MDIO driver for Thunder SoC's

This MDIO driver was already being built, but not installed due
to being selected by the ThunderX Ethernet driver.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: synchronize PCIE related options with arm64 defconfig

This turns on various PCI related options which are enabled
in the Linux kernel arch/arm64/configs/defconfig but not
yet in the OpenWrt config.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
armsr: armv8: enable AHCI/SATA controllers for mvebu,qoriq,juno

When comparing the generated OpenWrt .config to the Linux arm64
defconfig, I noticed these SATA controllers were not included.
As they may be used as a boot drive, they should be built into
the kernel.

CONFIG_SATA_MVEBU is for Marvell platforms.
CONFIG_SATA_QORIQ is for NXP Layerscape.
CONFIG_SATA_SIL24 is for Arm's Juno development board, see Linux
kernel commit d7c38ff1cd86 ("arm64: defconfig: Add Juno SATA
controller").

Signed-off-by: Mathew McBride <matt@traverse.com.au>

hauke commented on target/linux/armsr/armv8/config-6.1:
```diff
@@ -22,33 +51,58 @@ CONFIG_ARCH_WANTS_NO_INSTR=y
+CONFIG_ARM64_ERRATUM_2441007=y
+CONFIG_ARM64_ERRATUM_2441009=y
+CONFIG_ARM64_ERRATUM_2457168=y
+CONFIG_ARM64_ERRATUM_2658417=y
```
Please activate these too:
```
ARM64_ERRATUM_2253138
ARM64_ERRATUM_2224489
ARM64_ERRATUM_2054223
ARM64_ERRATUM_2067961
```

hauke replied:
It looks like these are not in the default configuration:
````
hauke@hauke-t480:~/linux/linux-next-clean$ git grep ARM64_ERRATUM_2253138
Documentation/arm64/silicon-errata.rst:| ARM            | Neoverse-N2     | #2253138        | ARM64_ERRATUM_2253138       |
arch/arm64/Kconfig:config ARM64_ERRATUM_2253138
arch/arm64/kernel/cpu_errata.c:#ifdef CONFIG_ARM64_ERRATUM_2253138
hauke@hauke-t480:~/linux/linux-next-clean$ git grep ARM64_ERRATUM_2224489
Documentation/arm64/silicon-errata.rst:| ARM            | Cortex-A710     | #2224489        | ARM64_ERRATUM_2224489       |
Documentation/arm64/silicon-errata.rst:| ARM            | Cortex-X2       | #2224489        | ARM64_ERRATUM_2224489       |
arch/arm64/Kconfig:config ARM64_ERRATUM_2224489
arch/arm64/kernel/cpu_errata.c:#ifdef CONFIG_ARM64_ERRATUM_2224489
hauke@hauke-t480:~/linux/linux-next-clean$ git grep ARM64_ERRATUM_2054223
Documentation/arm64/silicon-errata.rst:| ARM            | Cortex-A710     | #2054223        | ARM64_ERRATUM_2054223       |
arch/arm64/Kconfig:config ARM64_ERRATUM_2054223
arch/arm64/kernel/cpu_errata.c:#ifdef CONFIG_ARM64_ERRATUM_2054223
hauke@hauke-t480:~/linux/linux-next-clean$ git grep ARM64_ERRATUM_2067961
Documentation/arm64/silicon-errata.rst:| ARM            | Neoverse-N2     | #2067961        | ARM64_ERRATUM_2067961       |
arch/arm64/Kconfig:config ARM64_ERRATUM_2067961
arch/arm64/kernel/cpu_errata.c:#ifdef CONFIG_ARM64_ERRATUM_2067961
hauke@hauke-t480:~/linux/linux-next-clean$ 
````
They were recently added, maybe they are not so important.

mcbridematt replied:
> ARM64_ERRATUM_2253138
> ARM64_ERRATUM_2224489
> ARM64_ERRATUM_2054223
> ARM64_ERRATUM_2067961

These four are all relate to `CORESIGHT_TRBE` (CoreSight Trace Buffer extension) which does not get enabled on the default config or OpenWrt. On OpenWrt, CoreSight itself is disabled in the generic target config: https://github.com/openwrt/openwrt/blob/main/target/linux/generic/config-6.1#L1062

mcbridematt committed:
armsr: armv8: fix invalid symbol value for FSL_ENETC_QOS

The kernel FSL_ENETC_QOS option is only a compile time
option, it does not result in a separate module being built.

Set it to 'y' to resolve a warning from the kernel compile:

.config:2654:warning: symbol value 'm' invalid for FSL_ENETC_QOS

Signed-off-by: Mathew McBride <matt@traverse.com.au>
Fixes: c3151b6f045 ("armvirt: 64: add support for other SystemReady-compatible vendors")

mcbridematt committed:
armsr: enable ACPI_BUTTON

A review of the generated OpenWrt kernel .config
vs the Linux arm64 defconfig showed that this
option was not being enabled, as it is disabled
in OpenWrt's generic config.

ACPI_BUTTON is needed to report and respond to
power button events, so it should be enabled.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt committed:
kernel: default ARM_PMU on for armsr/armv8

CONFIG_ARM_PMU (Arm Performance Monitor Unit) is a requirement
to use KVM (virtualization) from Linux 5.11+, as the virtualised
guest has virtualized PMU access.

Signed-off-by: Mathew McBride <matt@traverse.com.au>

mcbridematt commented:
A couple more changes added:

* Default `CONFIG_ARM_PMU=y` (Arm Performance Monitor Unit) to on for armsr/armv8. Without PMU enabled, it is not possible to launch guest VMs using KVM.
  The change has to be done inside `Config-kernel.in` as this appears to override any setting in the target config-X file.
* Enabled ACPI Button driver
* Enabled AHCI/SATA for relevant platforms (mvebu, layerscape, arm juno dev board)
* Enabled PCI(e)-related options that were not in the OpenWrt configuration
* Fixed a kernel configuration warning due to `CONFIG_FSL_ENETC_QOS` not being a kernel module.

For reference, this is the current diff between Linux 6.1.45 arm64 defconfig and the config generated after this change:

https://gist.github.com/mcbridematt/ec4a0ec782a650927aa1a897e3b29440

(The erratas that are currently turned off relate to 32-bit compatibility, e.g only appear if CONFIG_COMPAT=y)
